File: correct-crypto/javascript/hmacHash.js
// Correct SHA-256 and HMAC-SHA-256 with constant-time verification (Node.js).
'use strict';
const crypto = require('crypto');

const sha256Hex = (data) => crypto.createHash('sha256').update(data).digest('hex');
const hmacSha256 = (key, data) => crypto.createHmac('sha256', key).update(data).digest();

function verify(key, data, tag) {
  const expected = hmacSha256(key, data);
  return expected.length === tag.length && crypto.timingSafeEqual(expected, tag);
}

const key = Buffer.from('secret-mac-key');
const tag = hmacSha256(key, Buffer.from('authenticate me'));
console.log('SHA-256:', sha256Hex('hello'));
console.log('HMAC-SHA-256 verify OK:', verify(key, Buffer.from('authenticate me'), tag));
module.exports = { sha256Hex, hmacSha256, verify };

File: correct-crypto/rust/hmac_hash.rs
// Correct SHA-256 hashing and HMAC-SHA-256 with constant-time verification.
use hmac::{Hmac, Mac};
use sha2::{Digest, Sha256};

type HmacSha256 = Hmac<Sha256>;

pub fn digest(data: &[u8]) -> [u8; 32] {
    let mut hasher = Sha256::new();
    hasher.update(data);
    hasher.finalize().into()
}

pub fn verify(key: &[u8], data: &[u8], tag: &[u8]) -> bool {
    let mut mac = HmacSha256::new_from_slice(key).expect("HMAC accepts any key length");
    mac.update(data);
    mac.verify_slice(tag).is_ok() // constant-time
}
